### DISCUSSION TITLE
后端跨域请求问题

### DISCUSSION THREAD

**Discussion by inkss** [ArtalkJS/Artalk#97] [Q&A] [answered]
```sh
➜  ArtalkGo ./artalk-go serve --config artalk-go.yml 
FATA[0000] 请检查配置文件，并设置一个 app_key (任意字符串) 用于数据加密 
```

```yml
# 加密密钥
app_key: "yeJyzdkj"
```

**Comment by qwqcode** [comment, marked as answer]
artalk-go.yml 配置文件和 artalk-go 程序是在同一个文件夹里吗

**Reply by inkss** [reply]
en

**Reply by inkss** [reply]
还有个问题啊，我现在pagekey写的是相对连接，不含域名，从控制台点进去的链接前面的域名成了api的域名了。

此外，使用导入功能，建议先上传，不要等返回。我用nginx反代访问的，响应过长nginx抛499了，看程序日志却是还在导。

**Reply by qwqcode** [reply]
PageKey 相对路径跳转问题已 fix https://github.com/ArtalkJS/Artalk/commit/e4eaaeb2be35d3059788584236023d076908fbc7
导入功能会在以后更新中改进，可以暂时使用命令行导入 `artalk-go help import`
参考：https://artalk.js.org/guide/transfer.html

**Reply by qwqcode** [reply]
后端 v2.1.0 增加指定工作目录的功能，使用参数 `-w`，例如：`artalk-go -w 工作目录 -c 配置文件` 这时配置文件路径使用相对路径，会基于工作目录，并且配置文件中的相对路径，例如数据库文件 './data/xxx' 也会基于工作目录。

如果 Artalk 还是无法找到配置文件，考虑使用绝对路径，或检查下文件是否存在。

**Reply by inkss** [reply]
其实我还碰到了各种花式cros问题，重新部署一版本，有问题再问你吧

**Reply by qwqcode** [reply]
你可能需要配置 trusted_domains
https://github.com/ArtalkJS/ArtalkGo/blob/master/artalk-go.example.yml#L28-L30
如果不是端口直接访问 ArtalkGo，例如 Nginx 反代之类的就是自己的问题了

**Reply by inkss** [reply]
![跨源资源共享（CORS） - HTTP  MDN 和另外 7 个页面 - 个人 - Microsof_22-04-15_21-37-39](https://user-images.githubusercontent.com/31947043/163577424-d746d260-ddde-4ab6-9670-48f961a09a63.png)


看了一些文档，真的迷迷糊糊的，没太看懂，预检请求不是浏览器发的吗，那就是我必须得设置这个响应头了？

**Reply by inkss** [reply]
另外一个题外话，我用fetch请求一个图片，资源是设置了`access-control-allow-headers: *` `access-control-allow-origin: *`，但是还是出现了cors，这是为什么呢，请求头里也没有含有身份凭证啊。

**Reply by qwqcode** [reply]
能否提供控制台的具体错误提示，我猜测可能是你页面是 https，而请求目标图片是 http 资源，所以才导致了这个问题

**Reply by inkss** [reply]
好像不是，还有这种，为什么因为图片没有设置`access-control-allow-headers`，然后请求失败呢：
```
No 'Access-Control-Allow-Origin' header is present on the requested resource.
```

**Reply by qwqcode** [reply]
你修改 trusted_domains 了吗

**Reply by inkss** [reply]
你来看吧，主要发生在登录后，各种报错。目前只是在本地测试，trusted_domains 写的是 `http://localhost`。

后端：`https://artalk.szyink.com/api`，密码：`test`；邮箱：`test@test.com`；密码：`test`。

部分 Nginx 配置：

```
proxy_pass http://127.0.0.1:8771;
add_header Access-Control-Allow-Origin "http://localhost";
add_header Access-Control-Allow-Methods "POST, GET, OPTIONS";
add_header Access-Control-Allow-Credentials true;
```

**Reply by qwqcode** [reply]
你把 trusted_domains 改成：

```
trusted_domains:
  '*'
```

**Reply by inkss** [reply]
附带身份凭证的请求的不是不能设置成*吗

**Reply by qwqcode** [reply]
如果本地 Artalk 前端不是默认端口，后端 trusted_domains 需要设置为例如： http://localhost:8080 http://localhost:8888 带上端口，而不是 http://localhost

**Reply by qwqcode** [reply]
如果实在解决不了，直接nginx配置

```
add_header 'Access-Control-Allow-Origin' '*' always;
add_header 'Access-Control-Allow-Credentials' 'true';
```

**Reply by inkss** [reply]
按照你说的，现在本地是没问题了

**Reply by inkss** [reply]
晚上部署的时候测测

**Reply by qwqcode** [reply]
> 还有个问题啊，我现在pagekey写的是相对连接，不含域名，从控制台点进去的链接前面的域名成了api的域名了。
> 
> 此外，使用导入功能，建议先上传，不要等返回。我用nginx反代访问的，响应过长nginx抛499了，看程序日志却是还在导。

后端 v2.1.1/0c69206 改为先上传了
